### PR TITLE
[machine][node] Move state out of the machine

### PR DESCRIPTION
**Pull request by cf19drofxots**
### Description

This is an initial PR to move the state handling out of the machine (as the machine is a stateless instruction executor), while introducing the node package which takes over the responsibility of handling said state.

- The machine now does not depend on the node package.
- This also introduces slightly better type safety around operation generation.

Remainder:
- [ ] update the iframe wallet to reflect these changes

Some of this will probably have to be refactored mainly due to:
- https://github.com/counterfactual/monorepo/pull/226
- https://github.com/counterfactual/monorepo/pull/163

### CODE REVIEW

File: packages/machine/src/instruction-executor.ts [hunk 1]
```diff
@@ -31,15 +31,15 @@ export class InstructionExecutor implements Observable {
...
+  public node: Node;
 
   // Observable
   public observers: Map<NotificationType, Function[]> = new Map();
 
   constructor(config: InstructionExecutorConfig) {
     this.responseHandler = config.responseHandler;
-    this.nodeState = new NodeState(config.state || {}, config.network);
-    this.middleware = new Middleware(this.nodeState, config.opGenerator);
+    this.node = new Node(config.state || {}, config.network);
```
Comment (snario) [inline packages/machine/src/instruction-executor.ts]: This seems like a bizarre decision, why would this class have a Node?

Comment (cf19drofxots) [reply packages/machine/src/instruction-executor.ts]: Handle to the context it's executing in.

Comment (snario) [reply packages/machine/src/instruction-executor.ts]: Can't we pass in only the minimum amount of information it needs? Like what does this class actually require to know?

Comment (cf19drofxots) [reply packages/machine/src/instruction-executor.ts]: We can, hence the WIP :)

File: .circleci/config.yml [hunk 2]
```diff
@@ -6,25 +6,26 @@ jobs:
     steps:
       - checkout
       - restore_cache:
-          key: dependency-cache-{{ checksum "packages/cf.js/package.json" }}{{ checksum "packages/contracts/package.json"}}{{ checksum "packages/iframe-wallet-poc/package.json"}}{{ checksum "packages/machine/package.json" }}{{  checksum "packages/playground/package.json" }}{{ checksum "packages/typescript-typings/package.json" }}
```
Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [inline .circleci/config.yml]: why change the key?

**Comment by IIIIllllIIIIllllIIIIllllIIIIllllIIIIll** [comment]
can you do a separate PR to rename `NodeState => Node` and `nodeState => node`. most of this PR is this change and it's hard to review.

**Comment by cf19drofxots** [comment]
@IIIIllllIIIIllllIIIIllllIIIIllllIIIIll ack - will do

File: .circleci/config.yml (hunk 2 above)
Comment (cf19drofxots) [reply .circleci/config.yml]: The PR introduces a new package so the key needs to be updated.

File: .circleci/config.yml [hunk 3]
```diff
@@ -6,25 +6,26 @@ jobs:
...
+          key: dependency-cache-{{ checksum "packages/cf.js/package.json" }}{{ checksum "packages/contracts/package.json"}}{{ checksum "packages/iframe-wallet-poc/package.json"}}{{ checksum "packages/machine/package.json" }}{{  checksum "packages/playground/package.json" }}{{ checksum "packages/typescript-typings/package.json" }}{{ checksum "packages/node/package.json" }}
       - run: yarn
       - save_cache:
-          key: dependency-cache-{{ checksum "packages/cf.js/package.json" }}{{ checksum "packages/contracts/package.json"}}{{ checksum "packages/iframe-wallet-poc/package.json"}}{{ checksum "packages/machine/package.json" }}{{  checksum "packages/playground/package.json" }}{{ checksum "packages/typescript-typings/package.json" }}
+          key: dependency-cache-{{ checksum "packages/cf.js/package.json" }}{{ checksum "packages/contracts/package.json"}}{{ checksum "packages/iframe-wallet-poc/package.json"}}{{ checksum "packages/machine/package.json" }}{{  checksum "packages/playground/package.json" }}{{ checksum "packages/typescript-typings/package.json" }}{{ checksum "packages/node/package.json" }}
```
Comment (snario) [inline .circleci/config.yml]: Does this thing support globs?

File: .circleci/config.yml [hunk 4]
```diff
@@ -6,25 +6,26 @@ jobs:
...
           paths:
             - node_modules
             - packages/contracts/node_modules
             - packages/iframe-wallet-poc/node_modules
             - packages/machine/node_modules
             - packages/cf.js/node_modules
             - packages/playground/node_modules
+            - packages/node/node_modules
```
Comment (snario) [inline .circleci/config.yml]: Also globs?

File: packages/cf.js/src/legacy/utils/index.ts [hunk 5]
```diff
@@ -2,7 +2,7 @@ import ethers from "ethers";
 import * as _ from "lodash";
 
 import * as abi from "./abi";
-import { FreeBalance } from "./free-balance";
+import { EMPTY_FREE_BALANCE, FreeBalance } from "./free-balance";
```
Comment (snario) [inline packages/cf.js/src/legacy/utils/index.ts]: Seems like a weird choice to export this like this. Shouldn't it be in a `constants` subdirectory like `ethers.constants`?

File: packages/machine/src/instruction-executor.ts [hunk 6]
```diff
@@ -1,11 +1,11 @@
 import * as cf from "@counterfactual/cf.js";
+import { Node } from "@counterfactual/node";
```
Comment (snario) [inline packages/machine/src/instruction-executor.ts]: This still seems like a massive antipattern to me. The `Node` should not be consumed by the `machine`. We should mark it as TO-FIX or something or just find a way to not do this in the first place.

File: packages/machine/src/middleware/state-transition/install-proposer.ts [hunk 7]
```diff
@@ -1,9 +1,9 @@
 import * as cf from "@counterfactual/cf.js";
+import { Node, StateChannelInfoImpl } from "@counterfactual/node";
```
Comment (snario) [inline packages/machine/src/middleware/state-transition/install-proposer.ts]: Also really dislike this dependency.

File: packages/node/package.json [hunk 8]
```diff
@@ -0,0 +1,53 @@
+{
+  "name": "@counterfactual/node",
+  "version": "0.0.1",
+  "main": "dist/index.js",
+  "types": "dist/src/index.d.ts",
+  "module": "dist/index.es.js",
+  "iife": "dist/index-iife.js",
```
Comment (snario) [inline packages/node/package.json]: Are all these various export methods actually necessary? We don't actually import this into the Playground environment yet so why don't we hold off on prematurely packaging it up until we need to?

File: packages/node/test/unit/channels.spec.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
+import * as cf from "@counterfactual/cf.js";
+import { ethers } from "ethers";
+import * as _ from "lodash";
+
+import { Node, StateChannelInfoImpl } from "../../src/node";
+
+describe("Node can contain channels", () => {
+  it("Node has no channels", () => {
```
Comment (snario) [inline packages/node/test/unit/channels.spec.ts]: Should write unit tests for basic instantiation e.g., https://github.com/gnosis/pm-js/blob/master/test/test_gnosis.js#L17-L35

File: packages/node/test/unit/channels.spec.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        cf.legacy.utils.EMPTY_FREE_BALANCE
+      )
+    };
+    const node = new Node(channels, cf.legacy.network.EMPTY_NETWORK_CONTEXT);
+    expect(_.size(node.channelStates)).toEqual(1);
+  });
+
+  it("Node has multiple channels", () => {
+    const channels = {
+      [ethers.Wallet.createRandom().address]: new StateChannelInfoImpl(
```
Comment (snario) [inline packages/node/test/unit/channels.spec.ts]: `ethers.Wallet.createRandom().address` => `ethers.utils.hexlify(ethers.utils.randomBytes(20))`

However as @mvanderh mentioned the other day, it's better to not literally compute random values each test.

File: packages/node/tsconfig.json [hunk 11]
```diff
@@ -0,0 +1,14 @@
+{
+  "extends": "../../tsconfig.json",
+  "compilerOptions": {
+    "rootDir": ".",
+    "sourceRoot": ".",
+    // TODO: Document why we need declarationDir here
+    "declarationDir": "dist/",
```
Comment (snario) [inline packages/node/tsconfig.json]: Do we need `declarationDir`?

File: packages/node/tsconfig.json [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+    "outDir": "dist/"
+  },
+  // TODO: Figure out how to add "test" here so they get linted by tslint when using
+  // the -p flag but *also* such that rollup doesn't include tests in the build.
```
Comment (snario) [inline packages/node/tsconfig.json]: If you remove `rollup` then we can get rid of this TODO.

File: packages/node/tsconfig.test.json [hunk 13]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (snario) [inline packages/node/tsconfig.test.json]: Do we need to have two of these?

File: packages/cf.js/src/legacy/utils/index.ts (hunk 5 above)
Comment (cf19drofxots) [reply packages/cf.js/src/legacy/utils/index.ts]: Yeah that's better.

File: packages/machine/src/instruction-executor.ts (hunk 6 above)
Comment (cf19drofxots) [reply packages/machine/src/instruction-executor.ts]: Not gonna mark as `FIXME` - rather not do it this way in the first place and introduce it to the codebase.

File: packages/machine/src/middleware/state-transition/install-proposer.ts (hunk 7 above)
Comment (cf19drofxots) [reply packages/machine/src/middleware/state-transition/install-proposer.ts]: Will also remove/refactor.

File: packages/node/test/unit/channels.spec.ts (hunk 10 above)
Comment (cf19drofxots) [reply packages/node/test/unit/channels.spec.ts]: I recall that as well. Generally 👍 to not compute random values each time.

File: packages/node/tsconfig.json (hunk 11 above)
Comment (cf19drofxots) [reply packages/node/tsconfig.json]: I do not think we do. No package consumes its typings.

File: packages/node/tsconfig.test.json (hunk 13 above)
Comment (cf19drofxots) [reply packages/node/tsconfig.test.json]: The only difference here is `"include": ["src", "test"],` which allows for compiling the tests, but not rolling up tests (since the `tsconfig.json` is used for rollup)

Comment (snario) [reply packages/node/tsconfig.test.json]: Seems like the rolls should be reversed then, e.g., `tsconfig.rollup.json` should be the edge case and `tsconfig.json` be the one that includes `src` and `test`

Comment (cf19drofxots) [reply packages/node/tsconfig.test.json]: That works too. Could either be `tsconfig.test.json` or `tsconfig.rollup.json`

I prefer your suggestion actually.

File: .circleci/config.yml (hunk 3 above)
Comment (cf19drofxots) [reply .circleci/config.yml]: Removed this formatting completely as CircleCI yaml parsing doesn't support interpreting globs.

https://github.com/counterfactual/monorepo/pull/235

**Comment by cf19drofxots** [comment]
The commits to setup the Node package have been cherry-picked into another PR: https://github.com/counterfactual/monorepo/pull/236/files

**Comment by cf19drofxots** [comment]
Closing this PR as 1) it's still blocked and 2) a lot of changes have occurred.

Will open another later when more appropriate.
